File: Solvers/NELDMD/NELDMD.m
%==========================================================================
%                       The Nelder-Mead Algorithm
%==========================================================================
% DATE
%        Feb 2016 (Updated in Dec 2019 by David Eckman)
%
% AUTHOR
%        Anna Dong, Nellie Wu

% REFERENCE		
%       Russell R. Barton, John S. Ivey, Jr., (1996)
%		Nelder-Mead Simplex Modifications for Simulation 
%		Optimization. Management Science 42(7):954-973.
%
%==========================================================================
%
% INPUT
%        problem
%              Problem function name
%        probstructHandle
%              Problem structure function name 
%        problemRng
%              Random number generators (streams) for problems
%        solverRng
%              Random number generator (stream) for solver
%
% OUTPUT
%        Ancalls
%              An array (size = 'NumSoln' X 1) of budget expended
%        A
%              An array (size = 'NumSoln' X 'dim') of solutions
%              returned by solver
%        AFnMean
%              An array (size = 'NumSoln' X 1) of estimates of expected
%              objective function value
%        AFnVar
%              An array of variances corresponding to
%              the objective function at A
%              Equals NaN if solution is infeasible
%        AFnGrad
%              An array of gradient estimates at A; not reported
%        AFnGradCov
%              An array of gradient covariance matrices at A; not reported
%        AConstraint
%              A vector of constraint function estimators; not applicable
%        AConstraintCov
%              An array of covariance matrices corresponding to the
%              constraint function at A; not applicable
%        AConstraintGrad
%              An array of constraint gradient estimators at A; not
%              applicable
%        AConstraintGradCov
%              An array of covariance matrices of constraint gradient
%              estimators at A; not applicable
%
%==========================================================================

%% Nelder-Mead
function [Ancalls, A, AFnMean, AFnVar, AFnGrad, AFnGradCov, ...
    AConstraint, AConstraintCov, AConstraintGrad, ...
    AConstraintGradCov] = NELDMD(probHandle, probstructHandle, problemRng, ...
    solverRng)

%% Unreported
AFnGrad = NaN;
AFnGradCov = NaN;
AConstraint = NaN;
AConstraintCov = NaN;
AConstraintGrad = NaN;
AConstraintGradCov = NaN;

% Separate the two solver random number streams
solverInitialRng = solverRng{1}; % RNG for finding initial solutions
solverInternalRng = solverRng{2}; % RNG for the solver's internal randomness
% Nelder-Mead does not use any internal randomness

% Set default values
r = 30; % Number of replications taken at each solution
alpha = 1;
gammap = 2;
betap = 0.5;
delta = 0.5;
sensitivity = 10^(-7); % shrinking scale for VarBds

% Get details of the problem
[minmax, dim, ~, ~, VarBds, ~, ~, ~, budget, ~, ~, ~] = probstructHandle(0);

% Shrink VarBds to prevent floating errors
VarBds(:,1) = VarBds(:,1) + sensitivity; 
VarBds(:,2) = VarBds(:,2) - sensitivity;

% Check for sufficiently large budget
numExtPts = dim + 1;
if budget < r*numExtPts % Need to evaluate all initial solns in ssolsM
    fprintf('A budget is too small for a good quality run of Nelder-Mead.');
    return
end

% Determine maximum number of solutions that can be sampled within max budget
MaxNumSoln = floor(budget/r); 

% Generate initial simplex from dim+1 random points
RandStream.setGlobalStream(solverInitialRng);
[~, ~, ~, ~, ~, ~, ~, ssolsM, ~, ~, ~] = probstructHandle(numExtPts);

% Initialize larger than necessary (extra point for end of budget)
Ancalls = zeros(MaxNumSoln, 1);
A = zeros(MaxNumSoln + 1, dim);
AFnMean = zeros(MaxNumSoln, 1);
AFnVar = zeros(MaxNumSoln, 1);

% Using CRN: for each solution, start at substream 1
problemseed = 1;
% If not using CRN: problemseed needs to be updated throughout the code

% Track overall budget spent
Bspent = 0;

%% Start Solving

% Evaluate solutions in initial structure 
fnV = zeros(numExtPts,1); % To track soln
fnVarV = zeros(numExtPts,1);
for i1 = 1:numExtPts
    [fn, FnVar, ~, ~, ~, ~, ~, ~] = probHandle(ssolsM(i1,:), r, problemRng, problemseed);
    Bspent = Bspent + r;
    fnV(i1) = -minmax*fn; % Minimize fn
    fnVarV(i1) = FnVar;
end

% Record initial solution data
Ancalls(1) = 0;
A(1,:) = ssolsM(1,:);
AFnMean(1) = -minmax*fnV(1); % flip sign back
AFnVar(1) = fnVarV(1);

% Sort solutions by obj function estimate
[l2hfnV,l2hfnIndV1] = sort(fnV);
l2hfnVarV = fnVarV(l2hfnIndV1,:);
ssolsMl2h = ssolsM(l2hfnIndV1,:);

% Record only when recommended solution changes
record_index = 2;

% Reflect Worst & Update ssolsMl2h
% Maximization problem is converted to minimization by -z.
while Bspent <= budget
    % Reflect worst point
    Phigh = ssolsMl2h(end,:); % Current worst pt
    Pcent = mean(ssolsMl2h(1:end-1,:)); % Centroid for other pts
    Prefl2 = Phigh; % Save the original point
    Prefl = (1+alpha)*Pcent - alpha*Phigh; % Reflection
    Prefl = checkCons(VarBds,Prefl,Prefl2); % Check if Prefl respects VarBds (if not, change it)        
    
    % Evaluate reflected point
    [Frefl, FreflVar, ~, ~, ~, ~, ~, ~] = probHandle(Prefl, r, problemRng, problemseed); % Cost r
    Bspent = Bspent + r;
    Frefl = -minmax*Frefl;
    
    % Track best, worst, and second worst points
    Plow = ssolsMl2h(1,:); % Current best pt
    Flow = l2hfnV(1);
    Fsechi = l2hfnV(end-1); % Current 2nd worst z
    Fhigh = l2hfnV(end); % Worst z from unreflected structure
    
    % Check if accept reflection
    if Flow <= Frefl && Frefl <= Fsechi
        
        ssolsMl2h(end,:) = Prefl; % Prefl replaces Phigh
        l2hfnV(end) = Frefl;
        l2hfnVarV(end) = FreflVar;
        
        % Sort & end updating
        [l2hfnV,l2hfnIndV] = sort(l2hfnV);
        l2hfnVarV = l2hfnVarV(l2hfnIndV,:);
        ssolsMl2h = ssolsMl2h(l2hfnIndV,:);
        
        % Best solution remains the same, so no reporting
        
    % Check if accept expansion (of reflection in the same direction)
    elseif Frefl < Flow
        
        Pexp2 = Prefl;
        Pexp = gammap*Prefl + (1-gammap)*Pcent;
        Pexp = checkCons(VarBds,Pexp,Pexp2);
        
        % Evaluate expansion point    
        [Fexp, FexpVar, ~, ~, ~, ~, ~, ~] = probHandle(Pexp, r, problemRng, problemseed); % Cost r
        Bspent = Bspent + r;        
        Fexp = -minmax*Fexp;
        
        % Check if expansion point is an improvement relative to simplex
        if Fexp < Flow
            
            % Pexp replaces Phigh
            ssolsMl2h(end,:) = Pexp; 
            l2hfnV(end) = Fexp;
            l2hfnVarV(end) = FexpVar;
            
            % Sort & end updating
            [l2hfnV,l2hfnIndV] = sort(l2hfnV);
            l2hfnVarV = l2hfnVarV(l2hfnIndV,:);
            ssolsMl2h = ssolsMl2h(l2hfnIndV,:);
            
            % Record data from expansion point (new best)
            if Bspent <= budget
                Ancalls(record_index) = Bspent;
                A(record_index,:) = Pexp;
                AFnMean(record_index) = -minmax*Fexp; % flip sign back
                AFnVar(record_index) = FexpVar;
                record_index = record_index + 1;
            end
            
        else
            
            % Prefl replaces Phigh
            ssolsMl2h(end,:) = Prefl; 
            l2hfnV(end) = Frefl;
            l2hfnVarV(end) = FreflVar;
            
            % Sort & end updating
            [l2hfnV,l2hfnIndV] = sort(l2hfnV);
            l2hfnVarV = l2hfnVarV(l2hfnIndV,:);
            ssolsMl2h = ssolsMl2h(l2hfnIndV,:);
            
            % Record data from reflected point (new best)
            if Bspent <= budget
                Ancalls(record_index) = Bspent;
                A(record_index,:) = Prefl;
                AFnMean(record_index) = -minmax*Frefl; % flip sign back
                AFnVar(record_index) = FreflVar;
                record_index = record_index + 1;
            end
            
        end
        
    % Check if accept contraction or shrink
    elseif Frefl > Fsechi % When Frefl is the worst z in nex simplex
        
        if Frefl <= Fhigh
            Phigh = Prefl; % Prefl replaces Phigh
            Fhigh = Frefl; % Frefl replaces Fhigh
        end
        
        % Attempt contraction or shrinking
        Pcont2 = Phigh;
        Pcont = betap*Phigh + (1-betap)*Pcent;
        Pcont = checkCons(VarBds,Pcont,Pcont2);
        
        % Evaluate contraction point
        [Fcont, FcontVar, ~, ~, ~, ~, ~, ~] = probHandle(Pcont, r, problemRng, problemseed);
        Bspent = Bspent + r;
        Fcont = -minmax*Fcont;
        
        % Accept contraction
        if Fcont <= Fhigh
            
            % Pcont replaces Phigh
            ssolsMl2h(end,:) = Pcont;
            l2hfnV(end) = Fcont;
            l2hfnVarV(end) = FcontVar;
            
            % Sort & end updating
            [l2hfnV,l2hfnIndV] = sort(l2hfnV);
            l2hfnVarV = l2hfnVarV(l2hfnIndV,:);
            ssolsMl2h = ssolsMl2h(l2hfnIndV,:);
            
            % Check if contraction point is new best
            if Fcont <= Flow
                % Record data from contraction point (new best)
                if Bspent <= budget
                    Ancalls(record_index) = Bspent;
                    A(record_index,:) = Pcont;
                    AFnMean(record_index) = -minmax*Fcont; % flip sign back
                    AFnVar(record_index) = FcontVar;
                    record_index = record_index + 1;
                end
            end
                
        else % Contraction fails -> Simplex shrinks by delta with Plow fixed
            
            ssolsMl2h(end,:) = Phigh; % Replaced by Prefl              
            
            % Check for new best
            new_best = 0;
            
            for i = 2:size(ssolsMl2h,1) % From Pseclo to Phigh
                Pnew2 = Plow;
                Pnew = delta*ssolsMl2h(i,:) + (1-delta)*Plow;
                Pnew = checkCons(VarBds,Pnew,Pnew2);
                [Fnew, FnewVar, ~, ~, ~, ~, ~, ~] = probHandle(Pnew, r, problemRng, problemseed); % Cost r/loop
                Bspent = Bspent + r;
                Fnew = -minmax*Fnew;
                
                % Check for new best
                if Fnew <= Flow
                    new_best = 1;
                end
                
                % Update ssolsM
                ssolsMl2h(i,:) = Pnew; % Pnew replaces Pi
                l2hfnV(i) = Fnew;
                l2hfnVarV(i) = FnewVar;
            end
                
            % Sort & end updating
            [l2hfnV,l2hfnIndV] = sort(l2hfnV);
            l2hfnVarV = l2hfnVarV(l2hfnIndV,:);
            ssolsMl2h = ssolsMl2h(l2hfnIndV,:);
            
            % Record data if there is a new best solution in the contraction
            if new_best == 1 && Bspent <= budget
                Ancalls(record_index) = Bspent;
                A(record_index,:) = ssolsMl2h(1,:);
                AFnMean(record_index) = -minmax*l2hfnV(1); % flip sign back
                AFnVar(record_index) = l2hfnVarV(1);
                record_index = record_index + 1;
            end
        end
    end
    
end

% Record solution at max budget
Ancalls(record_index) = budget;
A(record_index,:) = A(record_index - 1,:);
AFnMean(record_index) = AFnMean(record_index - 1);
AFnVar(record_index) = AFnVar(record_index - 1);

% Trim empty rows from data
Ancalls = Ancalls(1:record_index);
A = A(1:record_index,:);
AFnMean = AFnMean(1:record_index);
AFnVar = AFnVar(1:record_index);

%% Helper Functions

% Helper: Check & Modify (if needed) the new point, based on VarBds.
    function modiSsolsV = checkCons(VarBds,ssolsV,ssolsV2)
        col = size(ssolsV,2);
        stepV = ssolsV - ssolsV2;
        % t>0 for the correct direction
        tmaxV = ones(2,col);
        uV = VarBds(stepV>0,2); uV = uV';
        lV = VarBds(stepV<0,1); lV = lV';
        if ~isempty(uV) %length(uV)> 0
            tmaxV(1,stepV>0) = (uV - ssolsV2(stepV>0)) ./ stepV(stepV>0);
        end
        if ~isempty(lV) %length(lV)>0
            tmaxV(2,stepV<0) = (lV - ssolsV2(stepV<0)) ./ stepV(stepV<0);
        end
        t = min(min(tmaxV));
        modiSsolsV = ssolsV2 + t*stepV;
        %rounding error, may remove
        for kc=1:col
            if abs(modiSsolsV(kc))<0.00000005
                modiSsolsV(kc) = 0;
            end
        end
    end


end

File: simopt/solvers/neldmd.py
"""Nelder-Mead Algorithm.

Nelder-Mead: An algorithm that maintains a simplex of points that moves around the
feasible region according to certain geometric operations: reflection, expansion,
contraction, and shrinking. A detailed description of the solver can be found
`here <https://simopt.readthedocs.io/en/latest/neldmd.html>`__.
"""

from __future__ import annotations

from collections.abc import Iterable
from typing import Annotated, ClassVar

import numpy as np
from pydantic import Field

from simopt.base import (
    ConstraintType,
    ObjectiveType,
    Problem,
    Solution,
    Solver,
    SolverConfig,
    VariableType,
)


class NelderMeadConfig(SolverConfig):
    """Configuration for Nelder-Mead solver."""

    r: Annotated[
        int,
        Field(
            default=30,
            gt=0,
            description="number of replications taken at each solution",
        ),
    ]
    alpha: Annotated[
        float, Field(default=1.0, gt=0, description="reflection coefficient > 0")
    ]
    gammap: Annotated[
        float, Field(default=2.0, gt=1, description="expansion coefficient > 1")
    ]
    betap: Annotated[
        float,
        Field(
            default=0.5,
            gt=0,
            lt=1,
            description="contraction coefficient > 0, < 1",
        ),
    ]
    delta: Annotated[
        float,
        Field(default=0.5, gt=0, lt=1, description="shrink factor > 0, < 1"),
    ]
    sensitivity: Annotated[
        float,
        Field(default=1e-7, gt=0, description="shrinking scale for bounds"),
    ]
    initial_spread: Annotated[
        float,
        Field(
            default=0.1,
            gt=0,
            description="fraction of distance between bounds used for initial points",
        ),
    ]


class NelderMead(Solver):
    """Nelder-Mead Algorithm.

    The Nelder-Mead algorithm, which maintains a simplex of points that moves around
    the feasible region according to certain geometric operations: reflection,
    expansion, contraction, and shrinking.
    """

    name: str = "NELDMD"
    config_class: ClassVar[type[SolverConfig]] = NelderMeadConfig
    class_name_abbr: ClassVar[str] = "NELDMD"
    class_name: ClassVar[str] = "Nelder-Mead"
    objective_type: ClassVar[ObjectiveType] = ObjectiveType.SINGLE
    constraint_type: ClassVar[ConstraintType] = ConstraintType.BOX
    variable_type: ClassVar[VariableType] = VariableType.CONTINUOUS
    gradient_needed: ClassVar[bool] = False

    # FIXME: fix typing on `sort_sol`
    def solve(self, problem: Problem) -> None:  # noqa: D102
        # Designate random number generator for random sampling.
        get_rand_soln_rng = self.rng_list[1]
        n_pts = problem.dim + 1

        # Check for sufficiently large budget.
        if self.budget.total < self.factors["r"] * n_pts:
            err_msg = "Budget is too small for a good quality run of Nelder-Mead."
            raise ValueError(err_msg)

        # Shrink variable bounds to avoid floating errors.
        if problem.lower_bounds and not np.all(np.isneginf(problem.lower_bounds)):
            self.lower_bounds = (
                np.array(problem.lower_bounds) + self.factors["sensitivity"]
            )
        else:
            self.lower_bounds = None

        if problem.upper_bounds and not np.all(np.isposinf(problem.upper_bounds)):
            self.upper_bounds = (
                np.array(problem.upper_bounds) - self.factors["sensitivity"]
            )
        else:
            self.upper_bounds = None

        # Initial dim + 1 points.
        sol = [self.create_new_solution(problem.factors["initial_solution"], problem)]

        if self.lower_bounds is None or self.upper_bounds is None:
            sol.extend(
                self.create_new_solution(
                    problem.get_random_solution(get_rand_soln_rng), problem
                )
                for _ in range(1, n_pts)
            )
        else:  # Restrict starting shape/location.
            initial_solution = np.array(
                problem.factors["initial_solution"], dtype=float
            )
            distances = (self.upper_bounds - self.lower_bounds) * self.factors[
                "initial_spread"
            ]

            # Generate new points
            new_pts = np.tile(initial_solution, (problem.dim, 1))
            new_pts[np.arange(problem.dim), np.arange(problem.dim)] += distances

            # Apply boundary conditions
            out_of_bounds = (new_pts > self.upper_bounds) | (
                new_pts < self.lower_bounds
            )
            if np.any(out_of_bounds):
                new_pts[out_of_bounds] -= 2 * distances

            # If still out of bounds, set to nearest bound
            out_of_bounds = (new_pts > self.upper_bounds) | (
                new_pts < self.lower_bounds
            )
            if np.any(out_of_bounds):
                new_pts[out_of_bounds] = np.where(
                    np.array(problem.minmax)[np.newaxis, :] == -1,
                    self.lower_bounds,
                    self.upper_bounds,
                )

            sol.extend(self.create_new_solution(pt, problem) for pt in new_pts)

        r = self.factors["r"]  # For increasing replications.

        # Start Solving.
        # Evaluate solutions in initial structure.
        for solution in sol:
            self.budget.request(r)
            problem.simulate(solution, r)

        # Record initial solution data.
        # FIXME: I think this might be wrong
        self.intermediate_budgets.append(0)
        self.recommended_solns.append(sol[0])
        # Sort solutions by obj function estimate.
        sort_sol = self._sort_and_end_update(problem, sol)

        # Maximization problem is converted to minimization by using minmax.
        while True:
            # Shrink towards best if out of bounds.
            while True:
                # Reflect worst and update sort_sol.
                p_high = sort_sol[-1]  # Current worst point. # pyrefly: ignore
                p_high_x = np.array(p_high.x)
                p_cent = np.mean(
                    [s.x for s in sort_sol[:-1]],  # pyrefly: ignore
                    axis=0,
                )
                p_refl = np.array(
                    (1 + self.factors["alpha"]) * p_cent
                    - self.factors["alpha"] * p_high_x
                )

                # Check if reflection point is within bounds.
                if np.equal(p_refl, self._check_const(p_refl, p_high_x)).all():
                    break

                sol_0_x = np.array(sort_sol[0].x)  # pyrefly: ignore
                for i in range(1, len(sort_sol)):
                    p_new = (
                        self.factors["delta"]
                        * np.array(sort_sol[i].x)  # pyrefly: ignore
                        + (1 - self.factors["delta"]) * sol_0_x
                    )
                    p_new = self._check_const(p_new, sol_0_x)
                    p_new = Solution(p_new, problem)
                    p_new.attach_rngs(rng_list=self.solution_progenitor_rngs, copy=True)
                    self.budget.request(r)
                    problem.simulate(p_new, r)

                    # Update sort_sol.
                    sort_sol[i] = p_new  # p_new replaces pi.

                # Sort & end updating.
                sort_sol = self._sort_and_end_update(
                    problem,
                    sort_sol,  # pyrefly: ignore
                )

            # Evaluate reflected point.
            p_refl = tuple(p_refl.tolist())
            p_refl = Solution(p_refl, problem)
            p_refl.attach_rngs(rng_list=self.solution_progenitor_rngs, copy=True)
            self.budget.request(r)
            problem.simulate(p_refl, r)
            np_minmax = np.array(problem.minmax)
            refl_fn_val = np_minmax * -p_refl.objectives_mean

            # Track best, worst, and second worst points.
            p_low = sort_sol[0]  # Current best pt. # pyrefly: ignore
            inv_minmax = np_minmax * -1
            fn_low = inv_minmax * sort_sol[0].objectives_mean  # pyrefly: ignore
            fn_sec = inv_minmax * sort_sol[-2].objectives_mean  # pyrefly: ignore
            fn_high = inv_minmax * sort_sol[-1].objectives_mean  # pyrefly: ignore

            # Check if accept reflection.
            if fn_low <= refl_fn_val and refl_fn_val <= fn_sec:
                # The new point replaces the previous worst.
                sort_sol[-1] = p_refl  # pyrefly: ignore
                # Sort & end updating.
                sort_sol = self._sort_and_end_update(
                    problem,
                    sort_sol,  # pyrefly: ignore
                )
                # Best solution remains the same, so no reporting.

            # Check if accept expansion (of reflection in the same direction).
            elif refl_fn_val < fn_low:
                p_exp = self.factors["gammap"] * np.array(p_refl.x) + (
                    1 - self.factors["gammap"]
                ) * np.array(p_cent)
                p_exp = self._check_const(p_exp, p_refl.x)

                # Evaluate expansion point.
                p_exp = Solution(p_exp, problem)
                p_exp.attach_rngs(rng_list=self.solution_progenitor_rngs, copy=True)
                self.budget.request(r)
                problem.simulate(p_exp, r)
                exp_fn_val = inv_minmax * p_exp.objectives_mean

                # Check if expansion point is an improvement relative to simplex.
                sort_sol[-1] = (  # pyrefly: ignore
                    p_exp if exp_fn_val < fn_low else p_refl
                )

                # Sort & end updating.
                sort_sol = self._sort_and_end_update(
                    problem,
                    sort_sol,  # pyrefly: ignore
                )

                # Record data if within budget.
                self.intermediate_budgets.append(self.budget.used)
                self.recommended_solns.append(p_exp if exp_fn_val < fn_low else p_refl)

            # Check if accept contraction or shrink.
            elif refl_fn_val > fn_sec:
                if refl_fn_val <= fn_high:
                    p_high = p_refl  # p_refl replaces p_high.
                    fn_high = refl_fn_val  # Replace fn_high.

                # Attempt contraction or shrinking.
                p_cont2 = p_high
                p_cont = self.factors["betap"] * np.array(p_high.x) + (
                    1 - self.factors["betap"]
                ) * np.array(p_cent)
                p_cont = self._check_const(p_cont, p_cont2.x)

                # Evaluate contraction point.
                p_cont = Solution(p_cont, problem)
                p_cont.attach_rngs(rng_list=self.solution_progenitor_rngs, copy=True)
                self.budget.request(r)
                problem.simulate(p_cont, r)
                cont_fn_val = inv_minmax * p_cont.objectives_mean

                # Accept contraction.
                if cont_fn_val <= fn_high:
                    sort_sol[-1] = p_cont  # p_cont replaces p_high. # pyrefly: ignore

                    # Sort & end updating.
                    sort_sol = self._sort_and_end_update(
                        problem,
                        sort_sol,  # pyrefly: ignore
                    )

                    # Check if contraction point is new best.
                    if cont_fn_val < fn_low:
                        # Record data from contraction point (new best).
                        self.intermediate_budgets.append(self.budget.used)
                        self.recommended_solns.append(p_cont)
                # Contraction fails -> simplex shrinks by delta with p_low fixed.
                else:
                    # Set pre-loop variables
                    sort_sol[-1] = p_high  # Replaced by p_refl. # pyrefly: ignore
                    is_new_best = False
                    p_low_x = np.array(p_low.x)
                    for i in range(1, len(sort_sol)):  # pyrefly: ignore
                        p_new = (
                            self.factors["delta"] * np.array(sort_sol[i].x)
                            + (1 - self.factors["delta"]) * p_low_x
                        )
                        p_new = self._check_const(p_new, p_low.x)

                        p_new = Solution(p_new, problem)
                        p_new.attach_rngs(
                            rng_list=self.solution_progenitor_rngs, copy=True
                        )
                        self.budget.request(r)
                        problem.simulate(p_new, r)
                        new_fn_val = inv_minmax * p_new.objectives_mean

                        # Check for new best.
                        if new_fn_val <= fn_low:
                            is_new_best = True

                        # Update sort_sol.
                        sort_sol[i] = p_new  # p_new replaces pi.

                    # Sort & end updating.
                    sort_sol = self._sort_and_end_update(
                        problem,
                        sort_sol,  # pyrefly: ignore
                    )

                    # Record data if there is a new best solution in the contraction.
                    if is_new_best:
                        self.intermediate_budgets.append(self.budget.used)
                        self.recommended_solns.append(sort_sol[0])

    def _sort_and_end_update(
        self, problem: Problem, sol: Iterable[Solution]
    ) -> list[Solution]:
        """Sort solutions by objective values, accounting for minimization/maximization.

        Args:
            problem (Problem): The simulation-optimization problem defining the
                objective direction (minimize or maximize).
            sol (Iterable[Solution]): An iterable of solutions to be sorted.

        Returns:
            list[Solution]: A list of solutions sorted according to their
                objective values.
        """
        minmax_array = np.array(problem.minmax)
        return sorted(
            sol,
            key=lambda s: np.dot(minmax_array, s.objectives_mean),
            reverse=True,
        )

    def _check_const(
        self, new_point: Iterable[float], reference_point: Iterable[float]
    ) -> tuple:
        """Adjust a point to ensure it remains within the specified bounds.

        Args:
            new_point (Iterable[float]): The proposed new point to be checked and
                adjusted if necessary.
            reference_point (Iterable[float]): The original reference point used to
                compute movement direction.

        Returns:
            tuple: The modified point that adheres to the given bounds.
        """
        # Make sure everything is a NumPy array
        new_point = np.array(new_point)
        reference_point = np.array(reference_point)
        # Create or compute the other variables we need
        step = new_point - reference_point
        tmin = 1

        # Apply bounding constraints using NumPy masks
        if self.upper_bounds is not None:
            mask = step > 0
            if np.any(mask):
                tmin = min(
                    tmin,
                    np.min(
                        (self.upper_bounds[mask] - reference_point[mask]) / step[mask]
                    ),
                )

        if self.lower_bounds is not None:
            mask = step < 0
            if np.any(mask):
                tmin = min(
                    tmin,
                    np.min(
                        (self.lower_bounds[mask] - reference_point[mask]) / step[mask]
                    ),
                )

        # Compute the modified point
        adjusted_point = reference_point + tmin * step

        # Remove rounding errors
        adjusted_point[np.abs(adjusted_point) < self.factors["sensitivity"]] = 0

        return tuple(adjusted_point.tolist())
